### PR TITLE
KAFKA-7091: AdminClient should handle FindCoordinatorResponse errors

### PR DESCRIPTION
**Pull request by omkreddy**
- Update KafkaAdminClient implementation to handle FindCoordinatorResponse errors
- Remove scala AdminClient usage from core and streams tests

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by omkreddy** [comment]
@ijuma Pls take a look.

cc @mjsax  removed scala admin client usage from  streams test. Pls take a look.

**Comment by mjsax** [comment]
Thanks @omkreddy! The change to the integration test LGTM. Cannot comment on the other changes.

**Comment by ijuma** [comment]
@hachikuji should take a look. Is this a 2.0 blocker?

**Review by hachikuji — COMMENTED** [review]
@omkreddy Thanks for the patch. I think it is also worth handling `COORDINATOR_LOAD_IN_PROGRESS` since we can just retry blindly. Perhaps we can also check this error code when handling `describeConsumerGroups` API?

**Comment by omkreddy** [comment]
@hachikuji  Thanks for the review. Addressed the review comments. Please take one more look.

**Comment by omkreddy** [comment]
retest this please

File: clients/src/main/java/org/apache/kafka/common/requests/FindCoordinatorResponse.java [hunk 1]
```diff
@@ -107,6 +107,14 @@ public int throttleTimeMs() {
         return throttleTimeMs;
     }
 
+    public boolean hasError() {
+        return this.error != Errors.NONE;
+    }
+
+    public boolean retriableError() {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/requests/FindCoordinatorResponse.java]: Do we need this? Could we just check if the exception is an instance of `RetriableException`?

File: clients/src/main/java/org/apache/kafka/common/requests/FindCoordinatorResponse.java [hunk 2]
```diff
@@ -107,6 +107,14 @@ public int throttleTimeMs() {
...
+        return error == Errors.COORDINATOR_NOT_AVAILABLE || error == Errors.COORDINATOR_LOAD_IN_PROGRESS;
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/requests/FindCoordinatorResponse.java]: Can we update the possible error codes at the top of this file? It does not include `COORDINATOR_LOAD_IN_PROGRESS` and it incorrectly lists `NOT_COORDINATOR`.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 3]
```diff
@@ -2610,6 +2609,13 @@ public ListConsumerGroupOffsetsResult listConsumerGroupOffsets(final String grou
             void handleResponse(AbstractResponse abstractResponse) {
                 final FindCoordinatorResponse response = (FindCoordinatorResponse) abstractResponse;
 
+                if (response.retriableError()) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: nit: a little vexing to see this logic repeated. Maybe we could have a helper?
```java
if (handleFindCoordinatorError(response, future))
  return;
```

**Review by hachikuji — COMMENTED** [review]
Thanks for the update. Left a few minor comments.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 3 above)
Comment (omkreddy) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: @hachikuji  Thanks for the review. Updated the PR.

File: core/src/test/scala/integration/kafka/api/SaslSslAdminClientIntegrationTest.scala [hunk 4]
```diff
@@ -18,6 +18,8 @@ import java.util
 import kafka.security.auth.{All, Allow, Alter, AlterConfigs, Authorizer, ClusterAction, Create, Delete, Deny, Describe, Group, Operation, PermissionType, SimpleAclAuthorizer, Topic, Acl => AuthAcl, Resource => AuthResource}
 import kafka.server.KafkaConfig
 import kafka.utils.{CoreUtils, JaasTestUtils, TestUtils}
+import kafka.utils.TestUtils._
```
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/SaslSslAdminClientIntegrationTest.scala:21]: nit: seems we didn't need this import?

**Review by hachikuji — COMMENTED** [review]
@omkreddy Thanks, the fix looks good. Is it possible to modify one of the existing unit tests so that we can directly verify that retries are handled correctly? That would protect us from regressions in the future.

**Comment by omkreddy** [comment]
retest this please

File: core/src/test/scala/integration/kafka/api/SaslSslAdminClientIntegrationTest.scala (hunk 4 above)
Comment (omkreddy) [reply core/src/test/scala/integration/kafka/api/SaslSslAdminClientIntegrationTest.scala:21]: above import is for using TestUtils.assertFutureExceptionTypeEquals method

**Comment by omkreddy** [comment]
@hachikuji Updated the existing KafkaAdminClientTest.testDeleteConsumerGroups to verify retries by mocking error responses. Does this look good to you?

**Review by hachikuji — APPROVED** [review]
Thanks, LGTM

**Pull request merged**